File: algorithms/0075.sort_colors/sort_colors.rs
use std::collections::HashMap;

impl Solution {
    pub fn sort_colors(nums: &mut Vec<i32>) {
        let mut colors = HashMap::new();
        for &num in nums.iter() {
            *colors.entry(num).or_insert(0) += 1;
        }
        let mut index = 0;
        for n in 0..3 {
            let &count = colors.get(&n).unwrap_or(&0);
            for _ in 0..count {
                nums[index] = n;
                index += 1;
            }
        }
    }
}

File: algorithms/0075.sort_colors/sort_colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int lo = 0, hi = nums.size() - 1;
        for (int i = 0; i < nums.size(); ++i) {
            while (i <= hi && nums[i] == 2) {
                swap(nums[i], nums[hi]);
                hi -= 1;
            }
            if (nums[i] == 0) {
                swap(nums[i], nums[lo]);
                lo += 1;
            }
        }
    }
};